File: algorithms/0003/main.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
  // 假设输入的只有 ASCII 码, 不初始化为 -1 是因为 undefined >= num => false
  const locations = Array.apply(null, { length: 256 })
  let maxLen = 0
  let left = 0
  for (let i = 0; i < s.length; i++) {
    const curChar = s[i]
    if (locations[curChar] >= left) {
      left = locations[curChar] + 1
    } else if (i + 1 - left > maxLen) {
      maxLen = i + 1 - left
    }
    locations[curChar] = i
  }
  return maxLen
}

module.exports = lengthOfLongestSubstring

File: algorithms/0003/main.go
package p0003

func lengthOfLongestSubstring(s string) int {
	locations := [256]int{} // 只有256长是因为假定输入的字符串只有ASCII字符
	for i := range locations {
		locations[i] = -1 // 先设置所有的字符都没有见过
	}

	maxLen, left := 0, 0

	for i := 0; i < len(s); i++ {
		if locations[s[i]] >= left {
			left = locations[s[i]] + 1 // 在s[left:i+1]中去除s[i]字符及其之前的部分
		} else if i+1-left > maxLen {
			// fmt.Println(s[left:i+1])
			maxLen = i + 1 - left
		}
		locations[s[i]] = i
	}

	return maxLen
}
